[CIR] Implement__builtin_va_arg

mmha opened the pull request:
Part of https://github.com/llvm/llvm-project/issues/153286.
Depends on https://github.com/llvm/llvm-project/pull/153819.

This patch adds support for __builtin_va_arg by adding the cir.va.arg operator. Unlike the incubator it doesn't depend on any target specific lowering (yet) but maps to llvm.va_arg.

andykaylor commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -3492,4 +3492,49 @@ def CIR_VAEndOp : CIR_Op<"va.end"> {
+    frontend is responsible for providing a `va_list` pointer that matches the
+    target representation.
+
+    Unless replaced by LoweringPrepare for the chosen target ABI this maps to
```
It's probably best to just omit this paragraph.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -382,3 +382,13 @@ void CIRGenFunction::emitVAStartEnd(mlir::Value argValue, bool isStart) {
     cir::VAEndOp::create(builder, argValue.getLoc(), argValue);
 }
+
+// FIXME(cir): This completely abstracts away the ABI with a generic CIR Op. We
```
I think this comment doesn't belong. Completely abstracting away the ABI details is what we want in CIR. At most, this should be a comment saying that the ABI details will be handled in LoweringPrepare.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp:
```diff
@@ -382,3 +382,13 @@ void CIRGenFunction::emitVAStartEnd(mlir::Value argValue, bool isStart) {
+
+// FIXME(cir): This completely abstracts away the ABI with a generic CIR Op. We
+// need to decide how to handle va_arg target-specific codegen.
+mlir::Value CIRGenFunction::emitVAArg(VAArgExpr *ve, Address &vaListAddr) {
```
The `vaListAddr` parameter isn't used, and it doesn't look like it will be. That does raise a question though. Classic codegen uses this parameter when emitting aggregate or complex va args to detect when the ABI didn't support the type for the arg. We won't detect that until a later lowering phase. Is that a problem? I don't think we have access to the same diagnostic reporting mechanism at that point.

bcardosolopes commented:
> Depends on #153819.

For making the review process a bit more convenient I suggest you use some stacked approach, see https://llvm.org/docs/GitHub.html#stacked-pull-requests

xlauko commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -384,6 +384,19 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
   }
 
+  mlir::Value VisitVAArgExpr(VAArgExpr *ve) {
+    QualType Ty = ve->getType();
```
```suggestion
    QualType ty = ve->getType();
```

xlauko commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -384,6 +384,19 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+    }
+
+    Address argValue = Address::invalid();
+    mlir::Value val = cgf.emitVAArg(ve, argValue);
+
+    return val;
```
```suggestion
    return cgf.emitVAArg(ve, argValue);
```

xlauko commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -3415,4 +3415,126 @@ def CIR_FAbsOp : CIR_UnaryFPToFPBuiltinOp<"fabs", "FAbsOp"> {
+// Variadic Operations
+//===----------------------------------------------------------------------===//
+
+def CIR_VAStartOp : CIR_Op<"va.start"> {
```
To be more consistent in naming:

```suggestion
def CIR_VAStartOp : CIR_Op<"va_start"> {
```

and update its uses elsewhere.

xlauko commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -3415,4 +3415,126 @@ def CIR_FAbsOp : CIR_UnaryFPToFPBuiltinOp<"fabs", "FAbsOp"> {
+  }];
+}
+
+def CIR_VAEndOp : CIR_Op<"va.end"> {
```
```suggestion
def CIR_VAEndOp : CIR_Op<"va_end"> {
```

and update its uses elsewhere.

xlauko commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -3415,4 +3415,126 @@ def CIR_FAbsOp : CIR_UnaryFPToFPBuiltinOp<"fabs", "FAbsOp"> {
+  }];
+}
+
+def CIR_VAArgOp : CIR_Op<"va.arg"> {
```
```suggestion
def CIR_VAArgOp : CIR_Op<"va_arg"> {
```

and update its uses elsewhere.

mmha committed:
Address review feedback

andykaylor reviewed (approved):
lgtm

mmha committed:
[CIR] Implement__builtin_va_arg

Part of #153286.
Depends on #153819.

This patch adds support for __builtin_va_arg by adding the cir.va.arg operator. Unlike the incubator it doesn't depend on any target specific lowering (yet) but maps to llvm.va_arg.
